src/perf/event_provider.cpp

```diff
@@ -0,0 +1,384 @@
+    PERF_EVENT_HW("ref-cycles", REF_CPU_CYCLES),
+#endif
+};
+
+static const lo2s::platform::CounterDescription SW_EVENT_TABLE[] = {
+    PERF_EVENT_SW("cpu-clock", CPU_CLOCK),
+    PERF_EVENT_SW("task-clock", TASK_CLOCK),
+    PERF_EVENT_SW("page-faults", PAGE_FAULTS),
+    PERF_EVENT_SW("context-switches", CONTEXT_SWITCHES),
+    PERF_EVENT_SW("cpu-migrations", CPU_MIGRATIONS),
+    PERF_EVENT_SW("page-faults-minor", PAGE_FAULTS_MIN),
+    PERF_EVENT_SW("page-faults-major", PAGE_FAULTS_MAJ),
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 33)
+    PERF_EVENT_SW("aligment-faults", ALIGNMENT_FAULTS),
+    PERF_EVENT_SW("emulation-faults", EMULATION_FAULTS),
+#endif
+    /* PERF_EVENT_SW("dummy", DUMMY), */
+};
+
+#define PERF_MAKE_CACHE_ID(id) (id)
+#define PERF_MAKE_CACHE_OP_ID(id) ((id) << 8)
+#define PERF_MAKE_CACHE_OP_RES_ID(id) ((id) << 16)
+
+template <typename T>
+struct string_to_id
+{
+    const char* name;
+    T id;
+};
+
+static constexpr string_to_id<perf_hw_cache_id> CACHE_NAME_TABLE[] = {
+    { "L1-dcache", PERF_COUNT_HW_CACHE_L1D }, { "L1-icache", PERF_COUNT_HW_CACHE_L1I },
+    { "LLC", PERF_COUNT_HW_CACHE_LL },        { "dTLB", PERF_COUNT_HW_CACHE_DTLB },
+    { "iTLB", PERF_COUNT_HW_CACHE_ITLB },     { "branch", PERF_COUNT_HW_CACHE_BPU },
+};
+
+static constexpr string_to_id<perf_hw_cache_op_id> CACHE_OPERATION_TABLE[] = {
+    { "load", PERF_COUNT_HW_CACHE_OP_READ },
+    { "store", PERF_COUNT_HW_CACHE_OP_WRITE },
+    { "prefetch", PERF_COUNT_HW_CACHE_OP_PREFETCH },
+};
+
+static constexpr string_to_id<perf_hw_cache_op_result_id> CACHE_OP_RESULT_TABLE[] = {
+    /* use plural suffix on access; perf userland tools use
+       "(load|store|prefetche)s" to mean (load|store|prefetch)-accesses
+       TODO: fix pluralization prefetch -> prefetch(e)s */
+    { "s", PERF_COUNT_HW_CACHE_RESULT_ACCESS },
+    { "-misses", PERF_COUNT_HW_CACHE_RESULT_MISS },
+};
+
+inline constexpr std::uint64_t make_cache_config(perf_hw_cache_id cache, perf_hw_cache_op_id op,
+                                                 perf_hw_cache_op_result_id op_result)
+{
+    return cache | (op << 8) | (op_result << 16);
+}
+
+template <std::size_t N, typename T>
+inline constexpr std::size_t array_size(T (&)[N])
+{
+    return N;
+}
+}
+
+namespace lo2s
+{
+namespace perf
+{
+
+static bool event_is_openable(const platform::CounterDescription& ev)
+{
+    struct perf_event_attr attr;
+    memset(&attr, 0, sizeof(attr));
+    attr.size = sizeof(attr);
+    attr.type = ev.type;
+    attr.config = ev.config;
+    attr.config1 = ev.config1;
+    attr.exclude_kernel = lo2s::config().exclude_kernel;
+
+    int fd = syscall(__NR_perf_event_open, &attr, 0, -1, -1, 0);
+    if (fd == -1)
+    {
+        switch (errno)
+        {
+        case ENOTSUP:
+            Log::warn() << "perf event not supported by the running kernel: " << ev.name;
+            break;
+        default:
+            Log::warn() << "perf event not available: " << ev.name;
```

Comment:
I think this is too common on healthy systems, a reduction to `debug` is probably in order. When requested, this will trigger higher level warnings anyway.